library/Test/Hspec/Expectations/Json.hs

```diff
@@ -43,6 +48,45 @@ import Test.Hspec.Expectations.Json.Internal
 -- >>> let printFailure (HUnitFailure _ r) = putStr $ formatFailureReason r
 -- >>> let catchFailure f = handle printFailure $ f >> putStrLn "<passed>"
 
+data MatchOptions = MatchOptions
+  { tolerateMissingFields :: Bool
+  , enforceArrayOrdering :: Bool
+  , treatNullAsMissing :: Bool
+  }
+
+defaultMatchOptions :: MatchOptions
+defaultMatchOptions =
+  MatchOptions
+    { tolerateMissingFields = True
+    , enforceArrayOrdering = False
+    , treatNullAsMissing = False
+    }
+
+-- | Compare two JSON values, with a useful diff
+shouldMatchJsonWith :: HasCallStack => MatchOptions -> Value -> Value -> IO ()
+shouldMatchJsonWith matchOptions a b = unless (a == b) $ assertBoolWithDiff (a' == b') (toText b) (toText a)
+ where
+  toText = toStrict . decodeUtf8 . encodePretty . normalizeScientific
+  (a', b') = runFilters matchOptions a b
+
+runFilters :: MatchOptions -> Value -> Value -> (Value, Value)
+runFilters MatchOptions {..} a b = (a', b')
+ where
+  applyFilters = appEndo . foldMap (\(doIf, f) -> Endo $ if doIf then f else id)
+  -- Filters applied to both `a` and `b`
+  baseFilters =
+    [ (treatNullAsMissing, filterNullFields)
+    , (not enforceArrayOrdering, sortJsonArrays)
+    ]
+  -- In order to tolerateMissingFields we need to also prune `a` against `b`
+  a' =
+    applyFilters
+      ( baseFilters
+          <> [(tolerateMissingFields, \x -> pruneJson (Superset x) (Subset b))]
+      )
+      a
+  b' = applyFilters baseFilters b
```

Comment:
Ooooo... `s/shouldMatchJsonWith/shouldBeNormalizedJson/` :exploding_head:

Reply:
```hs
shouldBeJson = shouldBeJsonNormalized mempty
```

:exploding_head: :exploding_head: :exploding_head: 

Ok, I'll stop now. Just love it when abstractions fall into place.

Reply:
Yeah that naming is tough. I think of it as a sort of expansion or backfilling, but also is it no longer a subsetting at all? And the behavior is localized to only objects within arrays, right? Not an object at any depth. All of this nuance makes naming all but impossible here. `subsetActualToExpectedAsPerAJ`.

Real talk, you since we're passing around things now, it could be more natural to do `subsetActualToExpected SomeArrayPruneStrategy` -- i.e. your prune-strategy argument moved up at this level.

Reply:
The `MatchOptions` approach could still have backwards compat fwiw, we'd just use the old algorithm when `!treatNullAsMissing && tolerateMissingFields` and new one for `treatNullAsMissing && tolerateMissingFields`

Reply:
> One wrinkle here is that `treatNullsAsMissing <> subsetActualToExpected OldAlgorithm` will never really work as the old algorithm fails on arrays with different shaped elements.

I would take that set of options to be useful for,

```
{ a: 1, b: 1, c: [{ x: 1, y: 1}] }
  matches
    { a: 1, b: 1, c: [{x: 1}], d: null }
```

- `subsetActualToExpected OldAlgorithm` ensures the `c` array matches, and
- `treatNullsAsMissing` ensures `d` is ignored

Telling by way of normalizing to ignore something on the `Expected` side is odd (you control that, just remove it) but seems like _exactly_ what we've gotten ourselves into with a bunch of tests asserting `null`s and wanting to drop them out of the implementation side entirely.

Reply:
> Telling by way of normalizing to ignore something on the Expected side is odd (you control that, just remove it)

This isn't enough though!  The following returns false for example:
```
{a:[{b:1},{c:1}] `shouldMatchJson` {a:[{c:1},{b:1}]}
```
There is no possible way to play with the expected there to make this pass (aside from putting the array in the same order to hit the strict `unless (a==b)` case)

Reply:
I see the misunderstanding. I was just replying you saying it "doesn't really work", which I took to mean it was redundant or totally broken by our strategy.

I know it's not enough, and that that's why we're here at all. I just didn't realize you meant "doesn't really work [for my needs]" instead of "doesn't really work [at all]". I was confused because you started "one wrinkle", which also inferred as "great, we've solved my need with the new function, but..."

Reply:
To be totally clear, in my head, in the library we'd have:

```hs
shouldBeJson = shouldBeJsonNormalized mempty

shouldMatchJson = shouldBeJsonNormalized
  $ treatNullsAsMissing <> ignoreArrayOrdering <> subsetActualToExpected MatchFirstElement

data SubsetArrayElementsVia
  = MatchFirstElement -- ^ Current approach
  | BackfillElements -- ^ Your approach that adds null keys
```

And in our code back in megarepo, we'd define and use

```hs
shouldMatchJsonSomething = shouldBeJsonNormalized
  $ treatNullsAsMissing <> ignoreArrayOrdering <> subsetActualToExpected BackfillElements
```

Reply:
**Oh sorry!** I get it now. The type is `{a, b}` and one value has `a:null` and one value has `b:null` so you get the total change in keys that way. Got it, carry on.

Reply:
> Oh sorry! I get it now. The type is {a, b} and one value has a:null and one value has b:null so you get the total change in keys element-to-element that way. Got it, carry on.
 
👍 Just so I am clear, you think it's okay to change our pruning algorithm to do this normalization step always?

Reply:
> change our **pruning algorithm** to do this normalization step always?

I don't, though we can keep talking about it if you feel differently.

```hs
shouldBeJson = shouldBeJsonNormalized mempty

shouldMatchJson = shouldBeJsonNormalized
  $ treatNullsAsMissing
  <> ignoreArrayOrdering
  <> subsetActualToExpected MatchFirstElement

data SubsetArrayElementsVia
  = MatchFirstElement -- ^ Current approach
  | BackfillElements -- ^ Your approach that adds null keys
```

This is what I picture and keeps the algorithm the same _with regards to the array element pruning_. That behavior change remains too niche for me to feel good about.

It does introduce `treatNullsAsMissing` by default, which I am OK with. The reason is that "we treat null and omitted as equal in matching" is easy to understand and explain and is what `treatNullsAsMissing` should do (prune off any keys with null values, recursively, on both sides) . This is different but related to what you were actually looking to solve (with `subsetActualToExpected BackfillElements`), but seems like a reasonable thing to have (by default) if that's what you prefer.

By my plan, we would still have to define and use our own,

```hs
shouldMatchJsonSomething = shouldBeJsonNormalized
  $ treatNullsAsMissing
  <> ignoreArrayOrdering
  <> subsetActualToExpected BackfillElements
```

In megarepo to actually get your stuff working.

Reply:
My real objection is that you can shoot yourself in the foot with this API, but I guess that is not worse than the status quo?
 
Personally I view this change as being the more natural way to prune objects in arrays. We only need to normalize the `Subset` side which means it will not add any new keys to the `Superset`, but avoids issues of not finding a key in the first element of the `Subset` when it exists in other elements. This feels like the more natural way that a heterogenous list of objects could occur in the first place, i.e. a null field was omitted.

I can't come up with a case where this new algorithm produces different results than the old one, aside from fixing the issue with heterogenous lists?

Reply:
> I can't come up with a case where this new algorithm produces different results than the old one, aside from fixing the issue with heterogenous lists?

OK, that's fair. I can't come up with any counter-examples of bad behavior. I've only been pointing at unknown-unknowns as the risk. I yield, and will Approve the PR as-is.

> We only need to normalize the `Subset` side which means it will not add any new keys to the `Superset`

I think this may have made something click... Let me step back, tell me where I'm wrong (again, lol):

1. In concrete usage, `Superset` is the HTTP response and `Subset` is the hand-written value
2. We're removing any keys from `Superset` that don't appear in `Subset`
3. Then we come to Arrays...

**Before/current**:

- Locate first element of corresponding array in `Subset`
- Prune all keys of `Superset` not in that element

**After/changed**:

- Normalize `Subset` to all elements of the array have same keys (adding `null`)
- Locate first element of corresponding array in `Subset`
- Prune all keys of `Superset` not in that element

Is this right? Described this way it sounds like we don't want to muddy the single prune step a "strategy". Can't "normalize array elements on `Actual`" be its own step?

```hs
treatNullsAsMissing
  <> ignoreArrayOrdering
  <> expandHeterogenousArrays -- new step
  <> subsetActualToExpected -- same as before
```

Reply:
> Is this right?

Yep!

`expandHeterogenousArrays` is an interesting idea! It feels like that would make the ordering of `Normalizer`s matter though? Maybe that's okay?

Reply:
> that would make the ordering of Normalizers matter though? Maybe that's okay?

Yeah, I think they can and do and that's fine. It doesn't violate any Semigroup laws and it seems clear enough that these are modification passes, so of course they can interact with each other, so order will matter.